### ISSUE TITLE
Missing headers when using the C++ library

### ISSUE THREAD

**Issue by sgiraudot** [tensorflow/tensorflow#42492]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Mint 19
- TensorFlow installed from (source or binary): source
- TensorFlow version: r2.3
- Python version: 3.6
- Bazel version (if compiling from source): 3.1.0
- GCC/Compiler version (if compiling from source): clang 6.0.0

**Describe the problem**

I'm having a hard time using the C++ library of TensorFlow. First, is there an official documentation about how to compile it? I couldn't find anything on the official TensorFlow website. I used to use the `contrib/cmake` folder but it was removed. I only manage to compile the C++ library by gathering information here and there and following issues, but it's really hard to tell what is the correct procedure.

Anyway, I did manage to compile it using bazel (see below), but then when I try to compile my C++ program, I run into a missing file problem:

```sh
In file included from /home/simon/programs/third_party/tensorflow/tensorflow/cc/client/client_session.h:24:
In file included from /home/simon/programs/third_party/tensorflow/tensorflow/cc/framework/ops.h:21:
In file included from /home/simon/programs/third_party/tensorflow/tensorflow/core/framework/tensor.h:23:
/home/simon/programs/third_party/tensorflow/tensorflow/core/framework/allocator.h:24:10: fatal error: 'absl/strings/string_view.h' file not found
#include "absl/strings/string_view.h"
```

I found some issues on the tracker, but all of them seem to concern TensorFlow 1.x. It's not clear to me why this file is not included (and if it should be?). Anyway, I downloaded _absl_ by hand and included the files on my project. Now I run on this other missing file error:

```sh
In file included from /home/simon/programs/third_party/tensorflow/tensorflow/cc/client/client_session.h:24:
In file included from /home/simon/programs/third_party/tensorflow/tensorflow/cc/framework/ops.h:21:
In file included from /home/simon/programs/third_party/tensorflow/tensorflow/core/framework/tensor.h:24:
/home/simon/programs/third_party/tensorflow/tensorflow/core/framework/tensor_shape.h:22:10: fatal error: 'tensorflow/core/framework/types.pb.h' file not found
#include "tensorflow/core/framework/types.pb.h"
```

From what I found in the issues, this is a file that is supposed to be generated by bazel.

**Provide the exact sequence of commands / steps that you executed before running into the problem**

I used `./configure` with all default options, and I don't have Cuda/GPU support.

Then I tried:

```sh
bazel build -j 4 -c opt --verbose_failures //tensorflow:libtensorflow_cc.so
```

and also:

```sh
bazel build -j 4 -c opt --verbose_failures  //tensorflow:tensorflow_cc
```

but both give me the same error when using the library after compilation.

**Comment by sgiraudot** [comment]
I'm sorry to see that this issue has not been addressed in 4 months. Just to be clear, this is not a criticism, I do understand that the C++ library is not a priority and that TF developers are already quite busy with the Python version. In the mean time, should I consider the C++ library as officially broken and move on?

**Comment by Lennard94** [comment]
Dear sgiraudot,
maybe this is of help for you.
I also needed a C++ library for tensorflow to apply already trained models. The compilation of the required libraries was, however, frustrating. As you mentioned, the built time on a normal machine is very long, which hampers a systematic approach to test different options. Luckily, I have a HPC available, where I can run the compilation within seconds, allowing a more systematic approach.

I tried to compile the tensorflow libraries (libtensorflow.so, libtensorflow_cc.so and libtensorflow_framework.so)  starting from r2.4.0 to r1.15, followed by linking the libraries to my C++ project. The compilation of the libraries was usually not an issue at all, and all versions compiled successfully, if the protocol from the official side is followed accurately. For the interested reader, this means: Install Go, Bazelisk, the suggested python version and the gcc compiler. However, for all libraries >r2.0.0, I ran into linking issues in the final stage. Specifically, the parameter list of the function Tensor::TensorShape was different with regard to the header file. However, I noticed also discrepancies between other functions, which I could avoid using in my C++ project. For Tensor::TensorShape, I unsuccessfully tried various alternatives.

The final built, which was successful, used the options as follows:
**System Information**
- Centos 7
- Tensorflow version: r1.15
- Python version: 3.6
- GCC Compiler: 6.3.0
- Bazel version: (Bazelisk/go did not work for this library for me, so I compiled this to from source) 
- Latest absl version at the time writing this post
- cmake >3.1, fftw > 3.3.3, gsl > 2. (for compilation of the C++ project)
- command for compilation: bazel build --jobs 192 --config=opt [...]

Reading your post, it looks like that the references to the header files might be an issue for you.
This also took me some time, however, including the paths in the cmake as follows worked for me:


include_directories(
/cluster/work/igc/boeseltl/tensorflow_2.4.0/tensorflow/bazel-bin/  
/cluster/work/igc/boeseltl/tensorflow_2.4.0/tensorflow/bazel-bin/tensorflow/
/cluster/work/igc/boeseltl/tensorflow_2.4.0/
/cluster/work/igc/boeseltl/tensorflow_2.4.0/tensorflow/bazel-tensorflow 
/cluster/work/igc/boeseltl/tensorflow_2.4.0/tensorflow/bazel-tensorflow/external/eigen_archive/
/cluster/work/igc/boeseltl/tensorflow_2.4.0/abseil-cpp/
/cluster/work/igc/boeseltl/tensorflow_2.4.0/protobuf/src)

Do not get confused by the directory name, I just copy pasted it from the cmake file.
Linking in the final stage is obviously done as

link_directories(/cluster/work/igc/boeseltl/tensorflow_2.4.0/tensorflow/bazel-bin/tensorflow)
and
target_link_libraries(${LOOP_VAR} ${EXTERNAL_LIBRARIES} tensorflow_cc tensorflow_framework)



Finally, I would also like to ask the developer team, what their future plans are for the C++ API. 
Best, and good luck with your issue
Lennard
